File: app/modules/cloud-backup/ios/CloudBackupModule.swift
import ExpoModulesCore

/// Expo bridge exposing the user's iCloud CloudKit private database to JS for a
/// zero-knowledge wallet backup.
///
/// JS passes an opaque, already-encrypted blob as a `Uint8Array` (Expo marshals
/// it to `Data`); this module only stores/fetches those bytes in the user's own
/// CloudKit private database. It never sees keys or seeds. Mirrors
/// `SecureKeystoreModule`.
///
/// NOTE (configured later): this requires the iCloud + CloudKit capability and a
/// container id to be added to the app's entitlements
/// (`com.apple.developer.icloud-container-identifiers` /
/// `com.apple.developer.icloud-services = CloudKit`). `CloudKitBackupStore`
/// uses `CKContainer.default()`, which resolves to `iCloud.$(CFBundleIdentifier)`
/// once the entitlement exists. We intentionally do NOT hard-fail at build time
/// if the entitlement is absent — calls simply error/return unavailable at
/// runtime until a human wires up the container (see report).
public class CloudBackupModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CloudBackup")

    AsyncFunction("isAvailable") { () async throws -> Bool in
      try await CloudKitBackupStore.isAvailable()
    }

    AsyncFunction("putBlob") { (id: String, bytes: Data) async throws in
      try await CloudKitBackupStore.put(id: id, blob: bytes)
    }

    AsyncFunction("getBlob") { (id: String) async throws -> Data? in
      try await CloudKitBackupStore.get(id: id)
    }

    AsyncFunction("deleteBlob") { (id: String) async throws in
      try await CloudKitBackupStore.delete(id: id)
    }
  }
}

File: app/modules/cloud-backup/android/src/main/java/expo/modules/cloudbackup/CloudBackupModule.kt
package expo.modules.cloudbackup

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Android counterpart of the iOS CloudKit cloud backup.
 *
 * Stores a small (<1 KB) opaque, already-encrypted blob in the user's own
 * Google Drive `appDataFolder` (the hidden, per-app folder only this app can
 * see). JS passes the blob as a `Uint8Array` (Expo marshals to `ByteArray`);
 * this module only moves bytes and never sees keys or seeds. API mirrors
 * `CloudBackupModule.swift`.
 *
 * - `putBlob` creates-or-updates a file named `id` in appDataFolder (idempotent).
 * - `getBlob` finds + downloads the file for `id`, or returns null if absent.
 * - `deleteBlob` removes the file for `id` (no-op if absent).
 * - `isAvailable` reflects whether a signed-in account with the Drive appData
 *   scope is available.
 *
 * NOTE (configured later): the Drive OAuth scope
 * (`DriveScopes.DRIVE_APPDATA`) and a signed-in `GoogleAccountCredential` must
 * be wired up by the app's Google Sign-In flow before these calls succeed.
 * Until then `isAvailable` returns false and the mutating calls throw — by
 * design we do not hard-fail at build time. See `DriveBackupStore`.
 */
class CloudBackupModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("CloudBackup")

    AsyncFunction("isAvailable") {
      DriveBackupStore.isAvailable(appContext.reactContext)
    }

    AsyncFunction("putBlob") { id: String, bytes: ByteArray ->
      val drive = store() ?: throw unavailable()
      drive.put(id, bytes)
    }

    AsyncFunction("getBlob") { id: String ->
      val drive = store() ?: throw unavailable()
      drive.get(id) // ByteArray or null
    }

    AsyncFunction("deleteBlob") { id: String ->
      val drive = store() ?: throw unavailable()
      drive.delete(id)
    }
  }

  private fun store(): DriveBackupStore? = DriveBackupStore.from(appContext.reactContext)

  private fun unavailable() =
    CodedException("ERR_CLOUD_BACKUP", "Google Drive is unavailable — sign in and grant the Drive appData scope", null)
}
